source/max-staleness/max-staleness.md

```diff
@@ -0,0 +1,520 @@
+what.)
+
+S2's staleness estimate is:
+
+```
+SMax.lastWriteDate - S.lastWriteDate + heartbeatFrequencyMS
+= 20 - 5 + 10
+= 25
+```
+
+## Estimating Staleness: Example of Worst-Case Accuracy With Idle Replica Set
+
+Consider a primary P and a secondary S, and a client with heartbeatFrequencyMS set to 500 ms. There is no clock skew.
+(Previous examples show that skew has no effect.)
+
+The primary has been idle for 10 seconds and writes a no-op to the oplog at time 50 (meaning 50 seconds past midnight),
+and again at time 60.
+
+Before the secondary can replicate the no-op at time 60, the client checks both servers. The primary reports its
+lastWriteDate is 60, the secondary reports 50.
+
+The client estimates S's staleness as:
+
+```
+(S.lastUpdateTime - S.lastWriteDate) - (P.lastUpdateTime - P.lastWriteDate) + heartbeatFrequencyMS
+= (60 - 50) - (60 - 60) + 0.5
+= 10.5
+```
+
+The same story as a table:
+
+| Clock | Event                 | S.lastUpdateTime | S.lastWriteDate | P.lastUpdateTime | P.lastWriteDate | S staleness |
+| ----- | --------------------- | ---------------- | --------------- | ---------------- | --------------- | ----------- |
+| 50    | Idle write            | 50               |                 | 50               |                 |             |
+| 60    | Idle write begins     | 60               |                 | 50               |                 |             |
+| 60    | Client checks P and S | 60               | 60              | 50               | 60              | 10.5        |
+| 60    | Idle write completes  | 60               |                 | 60               |                 |             |
+
+> from table import Table
+>
+> data = \[\
+> \['Clock', 'Event', 'S.lastUpdateTime', 'S.lastWriteDate',\
+> 'P.lastUpdateTime', 'P.lastWriteDate', 'S
+> staleness'\],
+>
+> \['50', 'Idle write', '50', '', '50', '', ''\], \['60', 'Idle write begins', '60', '', '50', '', ''\], \['60', 'Client
+> checks P and S', '60', '60', '50', '60', '10.5'\], \['60', 'Idle write completes', '60', '', '60', '', ''\],
+>
+> \]
+>
+> print Table(data).create_table()
+
+In this scenario the actual secondary lag is between 0 and 10 seconds. But the staleness estimate can be as large as:
+
+```
+staleness = idleWritePeriodMS + heartbeatFrequencyMS
+```
+
+To ensure the secondary is always eligible for reads in an idle replica set, we require:
+
+```
+maxStalenessSeconds * 1000 >= heartbeatFrequencyMS + idleWritePeriodMS
+```
+
+## Supplemental
+
+Python scripts in this document's source directory:
+
+- \`test_max_staleness_spo.py\`: Uses `scipy.optimize` to determine worst-case accuracy of the staleness estimate in an
+  idle replica set.
+- \`test_staleness_estimate.py\`: Tests whether a client would correctly select a secondary from an idle replica set,
+  given a random distribution of values for maxStalenessSeconds, heartbeatFrequencyMS, lastWriteDate, and
+  lastUpdateTime.
```

Comment:
The \` does not render correctly in MD, this doesn't seem intentional.

https://github.com/mongodb/specifications/blob/39add29e0a3b6721688845874383878b90811a12/source/max-staleness/max-staleness.rst?plain=1#L374

```suggestion
- `test_max_staleness_spo.py`: Uses `scipy.optimize` to determine worst-case accuracy of the staleness estimate in an
  idle replica set.
- `test_staleness_estimate.py`: Tests whether a client would correctly select a secondary from an idle replica set,
  given a random distribution of values for maxStalenessSeconds, heartbeatFrequencyMS, lastWriteDate, and
  lastUpdateTime.
```